## 1. The oldest business in the world
<p><img src="https://assets.datacamp.com/production/repositories/5851/datasets/7dded924c6dc418d4a828f2f4daba99953c27a5a/400px-Eingang_zum_St._Peter_Stiftskeller.jpg" alt="The entrance to St. Peter Stiftskeller, a restaurant in Saltzburg, Austria. The sign over the entrance shows &quot;803&quot; - the year the business opened."></p>
<p><em>Image: St. Peter Stiftskeller, founded 803. Credit: <a href="https://commons.wikimedia.org/wiki/File:Eingang_zum_St._Peter_Stiftskeller.jpg">Pakeha</a>.</em></p>
<p>An important part of business is planning for the future and ensuring that the company survives changing market conditions. Some businesses do this really well and last for hundreds of years.</p>
<p>BusinessFinancing.co.uk <a href="https://businessfinancing.co.uk/the-oldest-company-in-almost-every-country">researched</a> the oldest company that is still in business in (almost) every country and compiled the results into a dataset. In this project, you'll explore that dataset to see what they found.</p>
<p>The database contains three tables.</p>
<h3 id="categories"><code>categories</code></h3>
<table>
<thead>
<tr>
<th style="text-align:left;">column</th>
<th>type</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;"><code>category_code</code></td>
<td>varchar</td>
<td>Code for the category of the business.</td>
</tr>
<tr>
<td style="text-align:left;"><code>category</code></td>
<td>varchar</td>
<td>Description of the business category.</td>
</tr>
</tbody>
</table>
<h3 id="countries"><code>countries</code></h3>
<table>
<thead>
<tr>
<th style="text-align:left;">column</th>
<th>type</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;"><code>country_code</code></td>
<td>varchar</td>
<td>ISO 3166-1 3-letter country code.</td>
</tr>
<tr>
<td style="text-align:left;"><code>country</code></td>
<td>varchar</td>
<td>Name of the country.</td>
</tr>
<tr>
<td style="text-align:left;"><code>continent</code></td>
<td>varchar</td>
<td>Name of the continent that the country exists in.</td>
</tr>
</tbody>
</table>
<h3 id="businesses"><code>businesses</code></h3>
<table>
<thead>
<tr>
<th style="text-align:left;">column</th>
<th>type</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;"><code>business</code></td>
<td>varchar</td>
<td>Name of the business.</td>
</tr>
<tr>
<td style="text-align:left;"><code>year_founded</code></td>
<td>int</td>
<td>Year the business was founded.</td>
</tr>
<tr>
<td style="text-align:left;"><code>category_code</code></td>
<td>varchar</td>
<td>Code for the category of the business.</td>
</tr>
<tr>
<td style="text-align:left;"><code>country_code</code></td>
<td>char</td>
<td>ISO 3166-1 3-letter country code.</td>
</tr>
</tbody>
</table>
<p>Let's begin by looking at the range of the founding years throughout the world.</p>

## configuration

Set up the required data for this notebook.

In [ ]:
# Install postgresql server
!sudo apt-get -y -qq update
!sudo apt-get -y -qq install postgresql
!sudo service postgresql start

# Setup a password `postgres` for username `postgres`
!sudo -u postgres psql -U postgres -c "ALTER USER postgres PASSWORD 'postgres';"

# Setup a database with name `tfio_demo` to be used
!sudo -u postgres psql -U postgres -c 'DROP DATABASE IF EXISTS tfio_demo;'
!sudo -u postgres psql -U postgres -c 'CREATE DATABASE tfio_demo;'

In [ ]:
!pip install --user --upgrade ipython-sql

In [2]:
%env DEMO_DATABASE_NAME=tfio_demo
%env DEMO_DATABASE_HOST=localhost
%env DEMO_DATABASE_PORT=5432
%env DEMO_DATABASE_USER=postgres
%env DEMO_DATABASE_PASS=postgres

env: DEMO_DATABASE_NAME=tfio_demo
env: DEMO_DATABASE_HOST=localhost
env: DEMO_DATABASE_PORT=5432
env: DEMO_DATABASE_USER=postgres
env: DEMO_DATABASE_PASS=postgres


In [3]:
!curl -s -OL https://raw.githubusercontent.com/cruseakshay/sql_practice/main/createdb.sql

!PGPASSWORD=$DEMO_DATABASE_PASS psql -q -h $DEMO_DATABASE_HOST -p $DEMO_DATABASE_PORT -U $DEMO_DATABASE_USER -d $DEMO_DATABASE_NAME -f createdb.sql

In [6]:
%load_ext sql

In [10]:
%%sql 
postgresql://postgres:postgres@localhost:5432/tfio_demo
 
-- Select the oldest and newest founding years from the businesses table
...

163 rows affected.


[('Hamoud Boualem', 1878, 'CAT11', 'DZA'),
 ('Communauté Électrique du Bénin', 1968, 'CAT10', 'BEN'),
 ('Botswana Meat Commission', 1965, 'CAT1', 'BWA'),
 ('Air Burkina', 1967, 'CAT2', 'BFA'),
 ('Brarudi', 1955, 'CAT9', 'BDI'),
 ('Cameroon Development Corporation', 1947, 'CAT1', 'CMR'),
 ('Correios de Cabo Verde', 1849, 'CAT16', 'CPV'),
 ('Banque Internationale pour la Centrafrique', 1946, 'CAT3', 'CAF'),
 ('Cotontchad', 1971, 'CAT1', 'TCD'),
 ('Central Bank of the Comoros', 1981, 'CAT3', 'COM'),
 ('Société nationale des Chemins de fer du Congo', 1889, 'CAT2', 'COD'),
 ('Development Bank of the Central African States', 1975, 'CAT3', 'COG'),
 ('La Poste', 1945, 'CAT16', 'CIV'),
 ('Ethio-Djibouti Railways', 1901, 'CAT2', 'DJI'),
 ('Egyptian National Railways', 1854, 'CAT2', 'EGY'),
 ('Guinea Ecuatorial Airlines', 1996, 'CAT2', 'GNQ'),
 ('Asmara Brewery', 1939, 'CAT9', 'ERI'),
 ('National Bank of Ethiopia', 1906, 'CAT3', 'ETH'),
 ('BGFIBank Group', 1971, 'CAT3', 'GAB'),
 ('Halco Mining', 1962, 'CAT15', 'GIN'),
 ('Correios da Guiné-Bissau', 1973, 'CAT16', 'GNB'),
 ('KCB Group Limited', 1896, 'CAT3', 'KEN'),
 ('Central Bank of Lesotho', 1978, 'CAT3', 'LSO'),
 ('National Port Authority', 1921, 'CAT2', 'LBR'),
 ('Umma Bank', 1907, 'CAT3', 'LBY'),
 ('Air Madagascar', 1962, 'CAT2', 'MDG'),
 ('Malawi Broadcasting Corporation', 1964, 'CAT13', 'MWI'),
 ('Office de Radiodiffusion-Télévision du Mali', 1957, 'CAT13', 'MLI'),
 ('Central Bank of Mauritania', 1973, 'CAT3', 'MRT'),
 ('Mauritius Post', 1772, 'CAT16', 'MUS'),
 ('Attijariwafa Bank', 1904, 'CAT3', 'MAR'),
 ('Beira Railroad Corporation', 1892, 'CAT2', 'MOZ'),
 ('NamPost', 1814, 'CAT16', 'NAM'),
 ('Office of Radio and Television of Niger', 1960, 'CAT13', 'NER'),
 ('First Bank of Nigeria', 1894, 'CAT3', 'NGA'),
 ('National Post Office', 1922, 'CAT16', 'RWA'),
 ('Central Bank of São Tomé and Príncipe', 1975, 'CAT3', 'STP'),
 ('Dakar–Niger Railway', 1924, 'CAT2', 'SEN'),
 ('Air Seychelles', 1977, 'CAT2', 'SYC'),
 ('Rokel Commercial Bank', 1917, 'CAT3', 'SLE'),
 ('Radio Mogadishu', 1943, 'CAT13', 'SOM'),
 ('Premier FMCG', 1820, 'CAT12', 'ZAF'),
 ('Ivory Bank', 1994, 'CAT3', 'SSD'),
 ('Bank of Khartoum', 1913, 'CAT3', 'SDN'),
 ('Swazi Rail', 1963, 'CAT2', 'SWZ'),
 ('Tanzania Breweries Limited', 1933, 'CAT9', 'TZA'),
 ('La Poste du Togo', 1883, 'CAT16', 'TGO'),
 ('La Poste Tunisienne', 1847, 'CAT16', 'TUN'),
 ('Stanbic Bank Uganda Limited', 1906, 'CAT3', 'UGA'),
 ('ZamPost', 1896, 'CAT16', 'ZMB'),
 ('Standard Chartered Zimbabwe', 1892, 'CAT3', 'ZWE'),
 ('Spinzar Cotton Company', 1930, 'CAT1', 'AFG'),
 ('Azerbaijan Caspian Shipping Company', 1858, 'CAT2', 'AZE'),
 ('BMMI', 1883, 'CAT17', 'BHR'),
 ('M. M. Ispahani Limited', 1820, 'CAT11', 'BGD'),
 ('Tashi Group', 1959, 'CAT2', 'BTN'),
 ('Hua Ho Department Store', 1947, 'CAT17', 'BRN'),
 ('National Bank of Cambodia', 1954, 'CAT3', 'KHM'),
 ("Ma Yu Ching's Bucket Chicken House", 1153, 'CAT4', 'CHN'),
 ('Bank of Georgia', 1903, 'CAT3', 'GEO'),
 ('Wadia Group', 1736, 'CAT12', 'IND'),
 ('Pindad', 1808, 'CAT8', 'IDN'),
 ('North Oil Company', 1928, 'CAT10', 'IRQ'),
 ('Café Abu Salem', 1914, 'CAT4', 'ISR'),
 ('Kongō Gumi', 578, 'CAT6', 'JPN'),
 ('Arab Bank', 1930, 'CAT3', 'JOR'),
 ('Bogatyr Access Komir', 1913, 'CAT15', 'KAZ'),
 ('M.H. Alshaya Co.', 1890, 'CAT17', 'KWT'),
 ('Electricite du Laos', 1959, 'CAT10', 'LAO'),
 ('Bank Audi', 1830, 'CAT3', 'LBN'),
 ('Pos Malaysia', 1800, 'CAT16', 'MYS'),
 ('Mongolian National Broadcaster', 1931, 'CAT13', 'MNG'),
 ('Myanmar National Airlines', 1948, 'CAT2', 'MMR'),
 ('Nepal Bank Limited', 1937, 'CAT3', 'NPL'),
 ("Kim Chong-t'ae Electric Locomotive Works", 1945, 'CAT2', 'PRK'),
 ('Petroleum Development Oman', 1937, 'CAT10', 'OMN'),
 ('House of Habib', 1841, 'CAT5', 'PAK'),
 ('Destileria Limtuaco', 1853, 'CAT9', 'PHL'),
 ('Salam International Investment Limited', 1952, 'CAT5', 'QAT'),
 ('Petrodvorets Watch Factory', 1721, 'CAT7', 'RUS'),
 ('House of Alireza', 1845, 'CAT6', 'SAU'),
 ('Singapore Post', 1819, 'CAT1

## 2. How many businesses were founded before 1000?
<p>Wow! That's a lot of variation between countries. In one country, the oldest business was only founded in 1999. By contrast, the oldest business in the world was founded back in 578. That's pretty incredible that a business has survived for more than a millennium.</p>
<p>I wonder how many other businesses there are like that.</p>

In [ ]:
%%sql

-- Get the count of rows in businesses where the founding year was before 1000
...

## 3. Which businesses were founded before 1000?
<p>Having a count is all very well, but I'd like more detail. Which businesses have been around for more than a millennium?</p>

In [ ]:
%%sql

-- Select all columns from businesses where the founding year was before 1000
-- Arrange the results from oldest to newest
...

## 4. Exploring the categories
<p>Now we know that the oldest, continuously operating company in the world is called Kongō Gumi. But was does that company do? The category codes in the <code>businesses</code> table aren't very helpful: the descriptions of the categories are stored in the <code>categories</code> table.</p>
<p>This is a common problem: for data storage, it's better to keep different types of data in different tables, but for analysis, you want all the data in one place. To solve this, you'll have to join the two tables together.</p>

In [ ]:
%%sql

-- Select business name, founding year, and country code from businesses; and category from categories
-- where the founding year was before 1000, arranged from oldest to newest
...

## 5. Counting the categories
<p>With that extra detail about the oldest businesses, we can see that Kongō Gumi is a construction company. In that list of six businesses, we also see a café, a winery, and a bar. The two companies recorded as "Manufacturing and Production" are both mints. That is, they produce currency.</p>
<p>I'm curious as to what other industries constitute the oldest companies around the world, and which industries are most common.</p>

In [ ]:
%%sql

-- Select the category and count of category (as "n")
-- arranged by descending count, limited to 10 most common categories
...

## 6. Oldest business by continent
<p>It looks like "Banking &amp; Finance" is the most popular category. Maybe that's where you should aim if you want to start a thousand-year business.</p>
<p>One thing we haven't looked at yet is where in the world these really old businesses are. To answer these questions, we'll need to join the <code>businesses</code> table to the <code>countries</code> table. Let's start by asking how old the oldest business is on each continent.</p>

In [ ]:
%%sql

-- Select the oldest founding year (as "oldest") from businesses, 
-- and continent from countries
-- for each continent, ordered from oldest to newest 
...

## 7. Joining everything for further analysis
<p>Interesting. There's a jump in time from the older businesses in Asia and Europe to the 16th Century oldest businesses in North and South America, then to the 18th and 19th Century oldest businesses in Africa and Oceania. </p>
<p>As mentioned earlier, when analyzing data it's often really helpful to have all the tables you want access to joined together into a single set of results that can be analyzed further. Here, that means we need to join all three tables.</p>

In [ ]:
%%sql

-- Select the business, founding year, category, country, and continent
...

## 8. Counting categories by continent
<p>Having <code>businesses</code> joined to <code>categories</code> and <code>countries</code> together means we can ask questions about both these things together. For example, which are the most common categories for the oldest businesses on each continent?</p>

In [ ]:
%%sql

-- Count the number of businesses in each continent and category
...

## 9. Filtering counts by continent and category
<p>Combining continent and business category led to a lot of results. It's difficult to see what is important. To trim this down to a manageable size, let's restrict the results to only continent/category pairs with a high count.</p>

In [ ]:
%%sql

-- Repeat that previous query, filtering for results having a count greater than 5
...